kythe/rust/fuchsia_extractor/README.md

```diff
@@ -1,9 +1,160 @@
 # Fuchsia Rust compilation extractor
```

Comment:
Can we be more precise with our wording here and elsewhere in this PR? This sounds like build system support for Fuchsia rather than an extractor that takes in a rust compilation invocation and outputs kzip data.

Is there a generic binary or library that does the extraction that is called/used by the Fuchsia build system support?

Reply:
> Can we be more precise with our wording here and elsewhere in this PR? This sounds like build system support for Fuchsia rather than an extractor that takes in a rust compilation invocation and outputs kzip data.

Not sure I understand the request here.  This is, indeed, a Fuchsia-specific compilation extractor for Rust. What would you rather see written here?

> Is there a generic binary or library that does the extraction that is called/used by the Fuchsia build system support?

I'm not sure what you are asking here.  I'll try to explain some background so maybe we can reformulate the question so it's easier for me to answer.

* The binary that extracts compilation is `rustc`, but invoked with a specific flag `-Zsave analysis`.  This program loads the output of `-Zsave-analysis` and converts into kzip files.
* However, since Fuchsia has a bespoke Rust toolchain, this specific program can not be applied to a piece of software other than Fuchsia. In that respect it is specifically support for Fuchsia's build system.
  * This, in turn, means that extraction can not be done "in parallel" to a regular build the way bazel extractors do it. Sigh.

Not sure if this helps.

Reply:
We talked about this in team chat and it sounds like rust makes the separation that we have in other supported languages difficult. For other languages we (try to) separate the extractor which is relatively generic and converts compiler invocations to kzip data from the support needed to get the build system to invoke the extractor. For example, there is build system support for gradle, bazel, cmake, etc and there are extractors for java, c++, go, etc. In some cases a light wrapper around the extractor is needed to support some build systems.

I was hoping that we could separate support for Fuchsia's build system from support for Rust in Fuchsia's build system so the build system support could be used for other languages as well and the core rust extractor could be used for other build systems.

Reply:
> We talked about this in team chat and it sounds like rust makes the separation that we have in other supported languages difficult. 

I see. I too am aware of the extractor/indexer separation of concerns that Kythe uses.  Ideally, Rust compilations should be separable this way.  However, *at the moment* this is expensive to do in Rust.  IIUC, work is underway in the Rust community to turn some of Rust internals into libraries, such that it would be possible to make a purely indexing pass, at which point we could have a generic Rust indexer.  However, today, if you wanted to replay a Rust compilation, you'd need pretty much to reimplement the `cargo` build system.

But until the said libraries are available, I think this is a pragmatic way forward. I think once they are, we'd be able to evolve the extractor and indexer to use them. (And even Kythe's current bazel extractor for Rust relies on the same mechanism, so I think we're aligned there.)

> I was hoping that we could separate support for Fuchsia's build system from support for Rust in Fuchsia's build system so the build system support could be used for other languages as well and the core rust extractor could be used for other build systems.

Similarly here, GN would need to support capturing compilations in sufficient detail to make this possible, which it currently does not. In Fuchsia, there's also ongoing work to make this happen, at which point we may be able to offer a generic tool.

So, even though the current approach could be thought of as suboptimal, it will give xref support for a few hundred Fuchsia engineers which should be a significant productivity boost.

Reply:
This is actually different than the extractor/indexer separation. This is build system/extractor separation. That is part of why I wanted to be more precise in the language, so we clearly document what is implemented. As you say this is a fine trade off for benefit vs complexity.

Reply:
Noted. I updated `README.md` to explain a bit, hopefully to your satisfaction. I'll resolve this discussion thread now, and if you want me to add specific things, please start a review and point out where you want things changed and how. Thanks!